frb_codegen/src/commands.rs

```diff
@@ -245,3 +250,111 @@ pub fn build_runner(dart_root: &str) -> Result {
     }
     Ok(())
 }
+
+lazy_static! {
+    static ref CACHE: Mutex<HashMap<String, String>> = Mutex::new(HashMap::new());
+}
+
+pub fn cargo_expand(dir: &str, module: Option<String>, file: &str) -> String {
+    let manifest_dir = env::var("CARGO_MANIFEST_DIR").unwrap_or_default();
+    #[cfg(windows)]
+    let manifest_dir = &manifest_dir.replace('\\', "/");
+    if !manifest_dir.is_empty() && dir == manifest_dir {
+        warn!(
+            "can not run cargo expand on {dir} because cargo is already running and would block cargo-expand. This might cause errors if your api contains macros."
+        );
+        return fs::read_to_string(file).unwrap_or_default();
+    }
+    let mut cache = CACHE.lock().unwrap();
+    let expanded = cache
+        .entry(String::from(dir))
+        .or_insert_with(|| run_cargo_expand(dir));
+    extract_module(expanded, module)
+}
+
+fn extract_module(mut expanded: &str, module: Option<String>) -> String {
+    if let Some(module) = module {
+        let mut spaces = 0;
+        for module in module.split("::") {
+            let searched = format!("mod {module} {{\n");
```

Comment:
currently we use string manipulation. shall we use token stream manipulation directly? (the downstream already parses string to token stream, so maybe we can move the parsing up to here)

Reply:
I don't think we can change that. In get_ir_file the source code is also used in text format.
```rust
    pub fn get_ir_file(&self) -> ParserResult<IrFile> {
        // info!("Phase: Parse source code to AST");
        let source_rust_content = read_rust_file(&PathBuf::from(&self.rust_input_path));
        let file_ast = syn::parse_file(&source_rust_content)?;

        // info!("Phase: Parse AST to IR");
        parser::parse(&source_rust_content, file_ast, &self.manifest_path)
    }

```

Reply:
Hmm you are right. I do hope it can be ast instead of string manipulation :)